tests/test_checkout.rs

```diff
@@ -73,6 +73,223 @@ fn test_checkout() {
+    Parent commit: 11c593341378 committed
+    Working copy : e6bed8446cf7 (no description set)
+    The working copy is clean
+    "###);
+}
+
+#[test]
+fn test_checkout_overwrite_empty_directory_with_normal_file() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_success(test_env.env_root(), &["init", "repo", "--git"]);
+    let repo_path = test_env.env_root().join("repo");
+    let victim_path = repo_path.join("file1");
+
+    // Commit normal file
+    std::fs::write(&victim_path, "committed\n").unwrap();
+    test_env.jj_cmd_success(&repo_path, &["describe", "-m", "committed"]);
+    test_env.jj_cmd_success(&repo_path, &["branch", "create", "committed"]);
+    test_env.jj_cmd_success(&repo_path, &["close"]);
+
+    // Create empty directory of the same name
+    test_env.jj_cmd_success(&repo_path, &["checkout", "root"]);
+    std::fs::create_dir(&victim_path).unwrap();
+
+    // Check out the committed file
+    // TODO: should remove the empty directory
+    let stderr = common::get_stderr_string(
+        &test_env
+            .jj_cmd(&repo_path, &["checkout", "committed"])
+            .assert(),
+    );
+    assert!(stderr.contains("Internal error: Failed to check out commit"));
+    /*
+    test_env.jj_cmd_success(&repo_path, &["checkout", "committed"]);
+    let stdout = test_env.jj_cmd_success(&repo_path, &["status"]);
+    insta::assert_snapshot!(stdout, @r###"
+    Parent commit: 11c593341378 committed
+    Working copy : 28012605e864 (no description set)
+    The working copy is clean
+    "###);
+    */
+}
+
+#[test]
+fn test_checkout_overwrite_non_empty_directory_with_normal_file() {
+    let test_env = TestEnvironment::default();
+    test_env.jj_cmd_success(test_env.env_root(), &["init", "repo", "--git"]);
+    let repo_path = test_env.env_root().join("repo");
+    let victim_path = repo_path.join("dir1");
+
+    // Commit normal file
+    std::fs::write(&victim_path, "committed\n").unwrap();
+    test_env.jj_cmd_success(&repo_path, &["describe", "-m", "committed"]);
+    test_env.jj_cmd_success(&repo_path, &["branch", "create", "committed"]);
+    test_env.jj_cmd_success(&repo_path, &["close"]);
+
+    // Create non-empty directory of the same name, but ignored
+    test_env.jj_cmd_success(&repo_path, &["checkout", "root"]);
+    std::fs::write(repo_path.join(".gitignore"), "dir1/\n").unwrap();
+    test_env.jj_cmd_success(&repo_path, &["describe", "-m", "ignore"]);
+    test_env.jj_cmd_success(&repo_path, &["new"]);
+    std::fs::create_dir(&victim_path).unwrap();
+    std::fs::write(victim_path.join("file1"), "ignored\n").unwrap();
+
+    let stdout = test_env.jj_cmd_success(&repo_path, &["status"]);
+    insta::assert_snapshot!(stdout, @r###"
+    Parent commit: 24c48bfe1eb4 ignore
+    Working copy : 47af6d92910e (no description set)
+    The working copy is clean
+    "###);
+
+    // Check out the committed file
+    // TODO: should we remove the ignored directory? or abort?
```

Comment:
Aborting is perhaps trickier than it seems because the transaction updating the repo has already committed at that point. We could still do that. It would mean that the working copy would be "stale" in the same way it is when you update another workspace's checkout. We currently automatically update the working copy to the desired commit when we notice that it's in such a stale state. That was just because it was the easiest way of handling the case, but I think it would be better to tell the user about the state and let them resolve it (e.g. by first removing the ignored directory or by updating back to a commit without the file).